[FX] Track use nodes in Node

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45775 [FX] Track use nodes in Node**

This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

Commit:
[FX] Track use nodes in Node

[ghstack-poisoned]

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

zdevito commented on torch/fx/node.py:
```diff
@@ -32,29 +32,10 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: Target,
-            map_arg(node.args, record_use)
-            map_arg(node.kwargs, record_use)
-        return use_nodes
+        # All of the nodes that use the value produced by this Node
```
assigning to `args` and `kwargs` also requires adjustment of the use lists. If you can't assign to them it makes it very hard to patch up the graph.

jamesr66a replied:
What do you think is a good solution? I could make them properties and automatically update `users` based on the difference of the old and new values. Is that good?

zdevito reviewed (commented):
I think this is missing some places where the use lists get invalidated. A follow up will need to support assigning to args, kwargs because otherwise it is hard to do pointwise surgery to the graph.

jamesr66a commented on torch/fx/node.py:
```diff
@@ -32,29 +32,10 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: Target,
+        # All of the nodes that use the value produced by this Node
+        # Note one user may correspond to several uses, e.g. the node fo `x + x`
+        # would appear once here, but represents two uses.
+        self.users : Set['Node'] = set()
```
What should the order here be? On insertion of a node, for example, the insertion point could be anywhere in the graph. Should we figure out where to insert user nodes to reflect the topological ordering? or seomthing else

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

jamesr66a commented on torch/fx/node.py:
```diff
@@ -64,18 +45,18 @@ def replace_all_uses_with(self, replace_with : 'Node') -> List['Node']:
                     return n
             use_node.args = map_arg(use_node.args, maybe_replace_node)
             use_node.kwargs = map_arg(use_node.kwargs, maybe_replace_node)
+            self.users.remove(use_node)
```
`replace_with` user updates should now be covered by `args` and `kwargs` property assignment

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

zdevito commented on torch/fx/node.py:
```diff
@@ -64,22 +78,23 @@ def replace_all_uses_with(self, replace_with : 'Node') -> List['Node']:
                     return replace_with
                 else:
                     return n
+
```
by assigning args/kwargs separately here there are two passes that have to fix up the uses might want to directly call `_update_kwargs_args`

zdevito commented on torch/fx/node.py:
```diff
@@ -64,22 +78,23 @@ def replace_all_uses_with(self, replace_with : 'Node') -> List['Node']:
             use_node.kwargs = new_kwargs
 
-        return use_nodes
+        self.users.clear()
```
with the current code this should be an assert right? because args/kwargs updates should have done it. Alternative that is more efficient would be to directly assign to `_args/_kwargs` above and then call clear here.

zdevito reviewed (approved):
Looks good!

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]

Commit:
Update on "[FX] Track use nodes in Node"


This turns `Node.uses` into a set that tracks the nodes that use `self`.

Differential Revision: [D24091082](https://our.internmc.facebook.com/intern/diff/D24091082)

[ghstack-poisoned]
